src/main.cpp

```diff
@@ -403,39 +403,287 @@ bool FullSyncWithDPORNodes()
+
+    // if dConfidence = 0, then the result must be 0.
+    if (!dConfidence)
+    {
+        if (fDebug10) LogPrintf("ETTS debug: Confidence of 0 specified: ETTS = %f", result);
+        return result;
+    }
 
-    return result/100;
+    bool staking = MinerStatus.nLastCoinStakeSearchInterval && MinerStatus.WeightSum;
+    // Get out early if not staking and set return value of 0.
+    if (!staking)
+    {
+        if (fDebug10) LogPrintf("ETTS debug: Not staking: ETTS = %f", result);
+        return result;
+    }
+
+    int64_t nValue = 0;
+    int64_t nCurrentTime = GetAdjustedTime();
+    if (fDebug10) LogPrintf("ETTS debug: nCurrentTime = %i", nCurrentTime);
+    
+    CTxDB txdb("r");
+    // Here I am defining a time mask 16 times as long as the normal stake time mask. This is to quantize the UTXO's into a maximum of
+    // 16 hours * 3600 / 256 = 225 time bins for evaluation. Otherwise for a large number of UTXO's, this algorithm could become
+    // really expensive.
+    const int ETTS_TIMESTAMP_MASK = 16 * STAKE_TIMESTAMP_MASK;
+
+    int64_t BalanceAvailForStaking = 0;
+    vector<COutput> vCoins;
+    
+    {
+        LOCK2(cs_main, pwalletMain->cs_wallet);
+        
+        BalanceAvailForStaking = pwalletMain->GetBalance() - nReserveBalance;
+
+        if (fDebug10) LogPrintf("ETTS debug: BalanceAvailForStaking = %u", BalanceAvailForStaking);
+        
+        // Get out early if no balance available and set return value of 0. This should already have happened above, because with no
+        // balance left after reserve, staking should be disabled; however, just to be safe...
+        if (BalanceAvailForStaking <= 0)
+        {
+            if (fDebug10) LogPrintf("ETTS debug: No balance available: ETTS = %f", result);
+            return result;
+        }
+
+        //reminder... void AvailableCoins(std::vector<COutput>& vCoins, bool fOnlyConfirmed=true, const CCoinControl *coinControl=NULL, bool fIncludeStakingCoins=false) const;
+        pwalletMain->AvailableCoins(vCoins, true, NULL, true);
+    }
+
+    
+    // An efficient local structure to store the UTXO's with the bare minimum info we need.
+    typedef vector< std::pair<int64_t, int64_t> > vCoinsExt;
+    vCoinsExt vUTXO;
+    // A local ordered set to store the unique "bins" corresponding to the UTXO transaction times. We are going to use this
+    // for the outer loop.
+    std::set<int64_t> UniqueUTXOTimes;
+        // We want the first "event" to be the CurrentTime. This does not have to be quantized.
+    UniqueUTXOTimes.insert(nCurrentTime);
+
+    
+    // Debug output cooldown...
+    if (fDebug10) LogPrintf("ETTS debug: nStakeMinAge = %i", nStakeMinAge);
+
+    // Derive a smoothed difficulty over desired PoSInterval of blocks by calling GetPoSKernelPS() which is netweight and reverse engineering...
+    double dDiff = GetAverageDifficulty(nPoSInterval);
+    if (fDebug10) LogPrintf("ETTS debug: dDiff = %f", dDiff);
```

Comment:
Allow this difficulty to be overridden by function parameter.

Reply:
Hmm... Both GetAverageDifficulty and GetEstimatedTimetoStake have default nPoSInterval arguments of 40. The intent here is to allow someone calling GetEstimatedTimetoStake a different choice of number of blocks back to look to average the Difficulty than the default 40. If I just called GetAverageDifficulty() here without the argument, it would be fixed at the 40, and I would not use the default argument nPoSInterval on GetEstimatedTimetoStake.

Reply:
I mean, I can see someone would want to see ETTS at arbitrary difficulty, not just the current. Instead of `nPoSInterval` and calculatiing the average difficulty internally, the function would take difficulty as argument.
`GetEstimatedTimetoStake(GetAverageDifficulty(80), 0.63)` or
`GetEstimatedTimetoStake(4, 0.70)`

Reply:
I get what you are saying now. Do you want me to change it and repush the commit? It is pretty easy...

Reply:
I do not think so, but you can do `double GetEstimatedTimetoStake(double dDifficulty = -1, double dConfidence = 0.8);`

Reply:
I did dDifficulty = 0.0, since normally specifying a dDiff of 0 is nonsensical. Makes the assertion for crazy parameters easier.